servant-auth-server/src/Servant/Auth/Server/Internal/Class.hs

```diff
@@ -1,49 +1,99 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
 {-# LANGUAGE UndecidableInstances #-}
 module Servant.Auth.Server.Internal.Class where
 
-import Servant.Auth
-import Data.Monoid
+import Data.Proxy
+import GHC.Generics (Generic)
+import Network.Wai (Request)
 import Servant hiding (BasicAuth)
 
+import Servant.Auth
 import Servant.Auth.Server.Internal.Types
 import Servant.Auth.Server.Internal.ConfigTypes
 import Servant.Auth.Server.Internal.BasicAuth
 import Servant.Auth.Server.Internal.Cookie
 import Servant.Auth.Server.Internal.JWT
 
+-- | The result of running the auths 'as'
+data AuthResult auths v
+  = Authenticated v
+  | AuthFailed (ErrorList (AuthsToErrors auths))
```

Comment:
Hm, does this need a third constructor for "hard" failures? E.g. if I make an API available under both Basic and JWT, and Basic is missing, JWT should be tried. If on the other hand Basic is invalid, should it maybe not just fail outright? (I actually am not sure what to think myself here.)

In any case, two things need to be distinguished somehow: auth being "indefinite" (e.g., there was no Basic Auth header) and it being "wrong" (bad username/password, invalid header). This is so we can provide the correct error message and status code.

Reply:
I had similar thoughts, and I can't really come up with a strong case either way.

Personally I'd lean towards leaving it as it is -- the server implementation can pattern match on the errors instead of on `AuthFailedHard` (or whatever the name would be), giving the same functionality. This requires concrete auth types though, so isn't possible for the test suites `server` function as it is currently written.

A downside to providing the extra constructor is that a `Maybe` (or similar) would need to be introduced in the error list (leading to something like `ErrorListMaybe auths`), since we aren't checking the other auth methods.